app/src/main/java/com/hedvig/app/feature/marketing/MarketingViewModel.kt

```diff
@@ -0,0 +1,240 @@
+        }
+    }
+
+    fun setLanguage(language: Language) {
+        val newState = updateMarketPickerState { it.copy(language = language) }
+        safeLet((newState as? PickMarket)?.market, (newState as? PickMarket)?.language) { m, l ->
+            updateApplicationLanguageUseCase.invoke(m, l)
+        }
+    }
+
+    private fun updateMarketPickerState(transform: (oldState: PickMarket) -> PickMarket): ViewState {
+        val oldState = state.value as? PickMarket
+            ?: throw IllegalStateException("Cannot update market picker state when not on market picker")
+        val newState = transform(oldState)
+        val newStateWithLanguage = newState.copy(
+            language = languageFromState(newState)
+        )
+        return _state.updateAndGet {
+            newStateWithLanguage.copy(
+                isValid = isValid(newStateWithLanguage),
+            )
+        }
+    }
+
+    private fun isValid(state: PickMarket): Boolean {
+        return (state.market != null && state.language != null)
+    }
+
+    private fun languageFromState(state: PickMarket): Language? {
+        val market = state.market ?: return state.language
+        val language = state.language ?: return defaultLanguage(market)
+
+        if (!isCompatibleLanguage(language, market)) {
+            return defaultLanguage(market)
+        }
+
+        return language
+    }
+
+    private fun isCompatibleLanguage(language: Language, market: Market) = when (market) {
+        Market.SE -> language == Language.EN_SE || language == Language.SV_SE
+        Market.NO -> language == Language.EN_NO || language == Language.NB_NO
+        Market.DK -> language == Language.EN_DK || language == Language.DA_DK
+        Market.FR -> language == Language.EN_FR || language == Language.FR_FR
+    }
+
+    private fun defaultLanguage(market: Market) = when (market) {
+        Market.SE -> Language.EN_SE
+        Market.NO -> Language.EN_NO
+        Market.DK -> Language.EN_DK
+        Market.FR -> Language.EN_FR
+    }
+
+    fun submitMarketAndLanguage() {
+        val currentState = state.value as? PickMarket ?: throw IllegalStateException("Can't submit, not in PickMarket")
+        if (!currentState.isValid) {
+            throw IllegalStateException("Can't submit, PickMarket is not valid")
+        }
+        val market = currentState.market ?: throw IllegalStateException("Can't submit, market is null")
+        val language = currentState.language ?: throw IllegalStateException("Can't submit, language is null")
+        updateMarketPickerState { it.copy(isLoading = true) }
+        viewModelScope.launch {
+            val result = submitMarketAndLanguagePreferencesUseCase
+                .invoke(language, market)
+            if (result.isLeft()) {
+                updateMarketPickerState { it.copy(isLoading = false) }
+                return@launch
+            }
+            updateApplicationLanguageUseCase.invoke(market, language)
```

Comment:
Do we need to update here, since we are calling this in `setMarket` and `setLanguage`?

Reply:
As it stands, we need to do it, in case the user does not explicitly make a selection and uses the values by default provided by the geolocation information.

Reply:
Ah I see